From c4ce705902522369d46e192a7bfe3a48503ea675 Mon Sep 17 00:00:00 2001
From: Zhang Xiao <xiao.zhang@windriver.com>
Date: Mon, 8 Jun 2009 14:06:12 +0800
Subject: [PATCH] Save rbp in interrupt stack when oprofile was configured

Signed-off-by: Zhang Xiao <xiao.zhang@windriver.com>

Common irq handler needn't rbp but oprofile needs it to backtrace
the user space stack. So just save the rbp when oprofile was configured.
---
 arch/x86/kernel/entry_64.S |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)

diff --git a/arch/x86/kernel/entry_64.S b/arch/x86/kernel/entry_64.S
index 38f92a0..13aa6a0 100644
--- a/arch/x86/kernel/entry_64.S
+++ b/arch/x86/kernel/entry_64.S
@@ -690,6 +690,12 @@ END(stub_rt_sigreturn)
 	SAVE_ARGS
 	leaq -ARGOFFSET(%rsp),%rdi	# arg1 for handler
 	pushq %rbp
+#if defined(CONFIG_OPROFILE) || defined(CONFIG_OPROFILE_MODULE)
+        /*
+         * Save the rbp to let oprofile backtrace the user stack
+        */
+        movq %rbp, -8(%rsp)
+#endif
 	CFI_ADJUST_CFA_OFFSET	8
 	CFI_REL_OFFSET		rbp, 0
 	movq %rsp,%rbp
-- 
1.6.3.3

